service/history/historyEngine.go

```diff
@@ -300,6 +300,25 @@ func (e *historyEngineImpl) GetWorkflowExecutionNextEventID(
 	}
 	defer release()
 
+	// if caller decide to long poll on workflow execution next event ID changes
+	if request.WaitForNewEvent != nil && *request.WaitForNewEvent == true {
+		subscribeID, channel, err := context.watchWorkflowExecution()
+		if err != nil {
+			return nil, err
+		}
+
+		context.Unlock()
+		timeoutChan := time.NewTimer(e.shard.GetConfig().LongPollExpirationInterval).C
```

Comment:
instead of wait on a fixed duration here, we should wait on the context (go's context.Context) of the service call.

Reply:
ok, add checking on ctx.Done()